art/attacks/poisoning/sleeper_agent_attack.py

```diff
@@ -0,0 +1,428 @@
+        return accuracy
+
+    # This function is responsible for returning indices of poison images with maximum gradient norm
+    @classmethod
+    def select_poison_indices(
+        cls, classifier: "CLASSIFIER_NEURALNETWORK_TYPE", x_samples: np.ndarray, y_samples: np.ndarray, num_poison: int
+    ) -> List[int]:
+        """
+        Select indices of poisoned samples
+
+        :classifier: Substitute Model.
+        :x_samples: Samples of poison.
+        :y_samples: Labels of samples of poison.
+        :num_poison: Number of poisoned samples to be selected out of all x_samples.
+        :return indices - Indices of samples to be poisoned.
+        """
+        import torch
+
+        device = "cuda" if torch.cuda.is_available() else "cpu"
+        grad_norms = []
+        criterion = torch.nn.CrossEntropyLoss()
+        model = classifier.model
+        model.eval()
+        differentiable_params = [p for p in classifier.model.parameters() if p.requires_grad]
+        for x, y in zip(x_samples, y_samples):
+            image = torch.tensor(x).to(device)
+            label = torch.tensor(y).to(device)
+            loss = criterion(model(image.unsqueeze(0)), label.unsqueeze(0))
+            gradients = torch.autograd.grad(loss, differentiable_params, only_inputs=True)
+            grad_norm = 0
+            for grad in gradients:
+                grad_norm += grad.detach().pow(2).sum()  # type: ignore
+            grad_norms.append(grad_norm.sqrt())  # type: ignore
+
+        indices = sorted(range(len(grad_norms)), key=lambda k: grad_norms[k])
+        indices = indices[-num_poison:]
+        return indices  # this will get only indices for target class
+
+    # This function is responsible for applying trigger patches to the images
+    # fixed - where the trigger is applied at the bottom right of the image
+    # random - where the trigger is applied at random location of the image
+    def apply_trigger_patch(self, x_trigger: np.ndarray) -> Tuple[Any]:
+        """
+        Select indices of poisoned samples
+
+        :x_trigger: Samples to be used for trigger.
+        :return tensor with applied trigger patches.
+        """
+        from art.estimators.classification.pytorch import PyTorchClassifier
+
+        patch_size = self.patch.shape[1]
+        if self.patching_strategy == "fixed":
+            x_trigger[:, -patch_size:, -patch_size:, :] = self.patch
+        else:
+            for x in x_trigger:
+                x_cord = random.randrange(0, x.shape[1] - self.patch.shape[1] + 1)
+                y_cord = random.randrange(0, x.shape[2] - self.patch.shape[2] + 1)
+                x[x_cord : x_cord + patch_size, y_cord : y_cord + patch_size, :] = self.patch
+        if isinstance(self.substitute_classifier, PyTorchClassifier):
+            import torch
+
+            return torch.tensor(np.transpose(x_trigger, [0, 3, 1, 2]))  # type: ignore
+
+        return np.transpose(x_trigger, [0, 3, 1, 2])  # type: ignore
```

Comment:
The transpose should only be necessary of if `self.estimator.channels_first == True`?